File: LeetCode/22.py
from typing import List

def generateParenthesis(n: int) -> List[str]:
    result = []

    def backtrace(s: str, open_count: int, closed_count: int):
        if len(s) >= 2*n:
            result.append(s)
            return
        
        if open_count < n:
            backtrace(s + "(", open_count + 1, closed_count)
        
        if closed_count < open_count:
            backtrace(s + ")", open_count, closed_count + 1)
        
    backtrace("", 0, 0)
    return result

print(generateParenthesis(3))

File: LeetCode/22.cpp
#include <iostream>
#include <vector>
#include <string>
#include <functional>

using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        std::function<void(string, int, int)> backtrack = [&backtrack, &result, &n](string s, int oc, int cc) {
            if (s.size() == 2 * n) {
                result.push_back(s);
                return;
            }

            if (oc < n) 
                backtrack(s + "(", oc + 1, cc);
            
            if (cc < oc) 
                backtrack(s + ")", oc, cc + 1);
        };
        backtrack("", 0, 0);
        return result;
    }
};

int main() {
    Solution task;
    int n;

    cin >> n;
    
    auto v = task.generateParenthesis(n);

    for(int i = 0; i < v.size(); i++) {
        cout << v[i] << endl;
    }

    return EXIT_SUCCESS;
}